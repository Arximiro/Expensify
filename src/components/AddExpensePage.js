import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);


// AddExpensePage is rendered by AppRouter which is rendered by App.js when the correct link is visted. Then, AddExpensePage renders ExpenseForm
// passing down via props the onSubmit function. Therefore, when the onSubmit event handler of the form in ExpenseForm is triggered, it runs this function.

// Connect is used to give access to dispatch. Dispatch is then used to dispatch the AddExpense action generator which adds the expense to the expenses array.

// Any component rendered by React-Router gets access to a lot of props, one of which is history, which has a bunch of methods.
// History has a function called push, which will redirect the user to whatever url is provided. This is used to redirect to the dashboard after adding an expense.

// mapDispatchToProps is used to make the component more testable, abstracting away the dispatch call.
// onSubmit is passed in via the props and it contains the dispatch call.
