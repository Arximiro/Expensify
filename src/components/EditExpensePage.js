import React from 'react';
import {connect} from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// We use mapStateToProps to get access to the current state of the expenses array, access provided by the (HOC) connect.
// React-Router renders the (HOC) thus providing acces to a bunch of other props, such as match.
// We use match to match the id of an expense in the expenses array to the id in the browser url, then return that expense.
// Then we pass the return value of mapStateToProps into EditExpensePage, allowing it to populate the expenseForm with
// the expense which we are wanting to edit.