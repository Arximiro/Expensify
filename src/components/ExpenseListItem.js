import React from 'react';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({id, description, note, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>Amount - {amount}</p>
        <p>Created At - {createdAt}</p>
        {note && <p>Note - {note}</p>}
    </div>
);

export default ExpenseListItem;

// Using connect in this manner will give us access to dispatch. Since we don't need access to state here we can just use connect
// to get access to what we need to which is the ability to dispatch.
