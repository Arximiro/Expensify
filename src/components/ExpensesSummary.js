import React from 'react';
import numeral from 'numeral';
import {connect} from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectVisibleExpenses from '../selectors/expenses';

export const ExpensesSummary = ({expenses, filters}) => {
    if (expenses.length > 0) {
        const visibleExpenes = selectVisibleExpenses(expenses, filters);
        const count = visibleExpenes.length;
        const total = selectExpensesTotal(visibleExpenes);
        const formattedTotal = numeral(total / 100).format('$0,0.00');
        const expenseText = count > 1 ? 'expenses' : 'expense';
        return (
        <p>Viewing {count} {expenseText} totalling {formattedTotal}</p>
        );
    } else {
        return (
            <p></p>
        );        
    }    
};

const mapStateToProps = (state) => ({expenses: state.expenses, filters: state.filters});

export default connect(mapStateToProps)(ExpensesSummary);
