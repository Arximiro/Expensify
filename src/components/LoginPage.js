import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startLoginGit, startLoginAnon } from '../actions/auth';

export const LoginPage = ({ startLogin, startLoginGit, startLoginAnon }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>It's time to get your expenses under control.</p>
            <button
                id="login1"
                className="button"
                onClick={startLogin}
            >
                Login with Google
            </button>
            <button
                id="login2"
                className="button"
                onClick={startLoginGit}
            >
                Login with Git
            </button>
            <button
                id="login3"
                className="button"
                onClick={startLoginAnon}
            >
                Login as Guest
            </button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
    startLoginGit: () => dispatch(startLoginGit()),
    startLoginAnon: () => dispatch(startLoginAnon())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
