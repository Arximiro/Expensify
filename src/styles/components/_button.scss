.button {
    background: rgb(113, 113, 231);
    border: none;
    border-radius: .5rem;
    color: white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    margin-bottom: $s-size;
    padding: $s-size;
    text-decoration: none;
    @media (min-width: $desktop-breakpoint) {        
        margin-right: $s-size;
        margin-bottom: none;
    }
}

.button--link {
    background: none;
}

.button--secondary {
    @extend .button;
    background: #888;
}
