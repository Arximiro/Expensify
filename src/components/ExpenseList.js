import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenes from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    {
        props.expenses.length === 0 ? (
            <p>No Expenses To Display</p>
        ) : (
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />
            })
        )
    }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenes(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

// Connect is a higher order component (HOC). It uses connect to connect us to the store via mapStateToProps. MSTP takes in the state and returns what we need
// from the state, in this component that is the expenses and filters. We pass pass those through the expenses selector to determine what should be visible.
// Connect then passes or *maps* the returned state to the props of ExpenseList, giving us access to the returned expenses.
// It's a component rendering a component, and in doing so passing the state of our store to that component in the form of props.
