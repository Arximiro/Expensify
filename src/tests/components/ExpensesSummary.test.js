import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/testExpenses';
import {filters} from '../fixtures/testFilters';

// --- Jest Notes ---
//

test('Should render ExpensesSummary with one expense', () => {
    const props = {
        expenses: [expenses[0]],
        filters
    };
    const res = ExpensesSummary(props);
    expect(res).toMatchSnapshot();
});

test('Should render ExpensesSummary with expenses', () => {
    const props = {
        expenses,
        filters
    };
    const res = ExpensesSummary(props);
    expect(res).toMatchSnapshot();
});
