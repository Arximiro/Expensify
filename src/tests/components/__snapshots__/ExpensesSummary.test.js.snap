// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with expenses 1`] = `
<p>
  Viewing 
  3
   
  expenses
   totalling 
  $1,043.00
</p>
`;

exports[`Should render ExpensesSummary with one expense 1`] = `
<p>
  Viewing 
  1
   
  expense
   totalling 
  $500.00
</p>
`;
